### PR TITLE
[R-package] fix OpenMP checking on macOS (fixes #4131)

### LINKED ISSUE #4131: [R Package] autoconf.ac OpenMP test ignores LDFLAGS (fix provided).
## Description
The LDFLAGS R variable "R CMD config LDFLAGS" is not included as part of the OpenMP configure.ac test in the R package.  This causes the compilation to not find OpenMP libraries due to ignoring the -L directories in .R/Makevars.  This is a larger issue on the Mac where -L is usually needed to find the OpenMP libraries.  Furthermore, since LDFLAGS are used for the package compilation, it would be a more reasonable test if LDFLAGS was used.

I wrote a fixed autoconf.ac, but I"m not sure how you would want to merge it due to the need to re-run autoconf.

## Reproducible example

From the Mac 

#### Current version does not look at the LDFLAGS when compiling the OpenMP test, therefore the test fails
/opt/homebrew/Cellar/llvm/11.1.0/bin/clang -o conftest conftest.c -lomp -Xclang -fopenmp

#### After adding the changes to autoconf.ac and running autoconf I was able to fix the issue and pass the OpenMP test by adding ${LDFLAGS} to the OpenMP test.

https://github.com/jlisic/LightGBM/commit/565dd56627b304392b21f205de2fa33069bc08cb#diff-9363f71bbcd7d64bb47ed62564d0bada0c8073646aaeecfe838068aa5bb17d2d

/opt/homebrew/Cellar/llvm/11.1.0/bin/clang -o conftest conftest.c -L/usr/local/lib -L/opt/homebrew/lib -lomp -L/opt/homebrew/opt/llvm/lib -Wl,-rpath,/opt/homebrew/opt/llvm/lib -lomp -Xclang -fopenmp

## Environment info

LightGBM version or commit hash: 3.2.0.99

M1  Mac running big-sur (R 4.0.3.). 
home-brew llvm 11.1.0 + libomp

cd lightgbm
R CMD INSTALL .

#### ~/.R/Makevars

CC=/opt/homebrew/Cellar/llvm/11.1.0/bin/clang
CXX=/opt/homebrew/Cellar/llvm/11.1.0/bin/clang++
CXX11=/opt/homebrew/Cellar/llvm/11.1.0/bin/clang++
CXX14=/opt/homebrew/Cellar/llvm/11.1.0/bin/clang++
CXX17=/opt/homebrew/Cellar/llvm/11.1.0/bin/clang++
CXX20=/opt/homebrew/Cellar/llvm/11.1.0/bin/clang++

CFLAGS=-g -O3 -Wall -pedantic -std=gnu99 -mtune=native -pipe  -fopenmp -I/opt/homebrew/include -I/opt/homebrew/opt/llvm/include
CXXFLAGS=-g -O3 -Wall -pedantic -std=c++11 -mtune=native -pipe -fopenmp  -I/opt/homebrew/include -I/opt/homebrew/opt/llvm/include
LDFLAGS=-L/usr/local/lib -L/opt/homebrew/lib -lomp -L/opt/homebrew/opt/llvm/lib -Wl,-rpath,/opt/homebrew/opt/llvm/lib

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #4131.

It makes a few improvements to the way that CRAN-style builds of the R package's `configure` script checks for the presence of OpenMP on macOS.

As a result, **OpenMP detection in the R package should be more reliable, meaning more users can benefit from multi-threading speedups**.

changes:

* passes the result of `R CMD config LDFLAGS` when running the OpenMP check, to ensure that that check uses the same linker flags that R itself will use when compiling `lib_lightgbm`
    - notably, this puts `-L/usr/local/lib` into the statement used to compile a test program, so brew-installed `libomp` will be found
    - this also allows users to install `{lightgbm}` with non-standard install locations for `libomp` by modifying `LDFLAGS` in `~/.R/Makevars`
* adds an additional check using only `-fopenmp` if compilation with `-Xclang -fopenmp` fails
    - this makes OpenMP detection work when compiling a CRAN package using `g++` (on `master`, it fails with a complaint that flag `-Xclang` is not recognized)
    - see https://stackoverflow.com/a/12002492/3986677
* compiles the test program with a C++ compiler instead of a C compiler

### Notes for Reviewers

CRAN prepares precompiled binaries for macOS, so this change should only affect people building from source using the repo, or choosing to do custom compilation from CRAN with something like `install.packages("lightgbm", type = "source")`.

As noted above, this should also help with detection of OpenMP when not using `clang` on Mac.

<details><summary>click here to learn how I tested using g++ on Mac</summary>

I have OpenMP at the following loccations on my laptop.

```shell
find /usr -name 'libomp.dylib'
```

```text
/usr/local/lib/libomp.dylib
/usr/local/Cellar/libomp/12.0.0/lib/libomp.dylib
```

My `~/.R/Makevars` contains only the following.

```shell
CXX=/usr/local/bin/g++-11
CXX11=/usr/local/bin/g++-11
CC=/usr/local/bin/gcc-11
```

When I build the CRAN package from source on `master`, OpenMP isn't found.

```shell
Rscript -e "remove.packages('lightgbm')"
sh build-cran-package.sh
R CMD INSTALL lightgbm_3.2.1.99.tar.gz > out.txt 2>&1
cat out.txt | grep -E "checking whether OpenMP"
```

> checking whether OpenMP will work in a package... no

On this PR, it's found.

</details>

### CODE REVIEW

**Comment by jameslamb** [comment]
The `valgrind` test is failing on complaints about a missing dependency

> * installing to library ‘/usr/local/RDvalgrind/lib/R/site-library’
ERROR: dependency ‘Matrix’ is not available for package ‘lightgbm’

PROBABLY not related to this PR's changes, but it'll be easy enough to test after #4503 is merged.

**Comment by jameslamb** [comment]
Solaris checks are failing with the following error

> ERROR: dependency ‘R6’ is not available for package ‘lightgbm’

and

> downloaded 59 KB
trying URL 'https://cloud.r-project.org/src/contrib/praise_1.0.0.tar.gz'
Error in download.file(url, destfile, method, mode = "wb", ...) :
cannot open URL 'https://cloud.r-project.org/src/contrib/praise_1.0.0.tar.gz'
In addition: Warning messages:
1: In download.file(url, destfile, method, mode = "wb", ...) :
downloaded length 0 != reported length 63
2: In download.file(url, destfile, method, mode = "wb", ...) :
cannot open URL 'https://cloud.r-project.org/src/contrib/praise_1.0.0.tar.gz': HTTP status was '503 Service Unavailable'
Warning in download.packages(pkgs, destdir = tmpd, available = available, :
download of package ‘praise’ failed

and

> Error: Bioconductor version cannot be validated; no internet connection?
In addition: Warning messages:
1: In file(con, "r") :
cannot open URL 'https://bioconductor.org/config.yaml': HTTP status was '503 Service Unavailable'
2: In file(con, "r") :
cannot open URL 'http://bioconductor.org/config.yaml': HTTP status was '503 Service Unavailable'
Execution halted
ERROR: lazy loading failed for package ‘R6’

and

> ERROR: dependency ‘R6’ is not available for package ‘processx’

<hr>

I'm going to run it again, and if we see similar failures I'll report them at https://github.com/r-hub/rhub.

**Comment by jameslamb** [comment]
> I'm going to run it again, and if we see similar failures I'll report them at https://github.com/r-hub/rhub.

I decided not to run these again, since other PRs will need to get merged before this one anyway (like #4506).

@StrikerRUS I see you're starting to merge things now that the blocking CI issue (#4524) is fixed. Don't worry about updating this PR. I can do that tonight, since it might require resolving conflicts with #4506.

**Review by StrikerRUS — APPROVED** [review]
Makes sense to me, LGTM!
Thanks a lot for great improvements in `configure.ac`!

File: R-package/configure
Comment (StrikerRUS) [inline R-package/configure:1828]: @jameslamb Sorry for commenting in closed PR.
I just thought about this a little bit.
This if branch is mostly for "not AppleClang compiler", right? Then why do we test only `-fopenmp` flag? Maybe it's worth checking some other flags, e.g. `-qopenmp` and `-openmp` for Intel compiler?
Possible values: https://github.com/autoconf-archive/autoconf-archive/blob/6c25681d78d963708362ea0d4d63ee9b508782b6/m4/ax_openmp.m4#L82

https://www.gnu.org/software/autoconf-archive/ax_openmp.html

Comment (jameslamb) [reply R-package/configure:1828]: oooo interesting! Yes, this branch is for "not clang", and was inspired by what I saw other `data.table` doing (since I know those maintainers have spent a lot of time struggling with OpenMP detection in R packages).

For example, 

https://github.com/Rdatatable/data.table/blob/4cf92893742a0359ce3b3ddfd279a9e69627d52b/configure#L91

```shell
# Compilation failed -- try forcing -fopenmp instead.
R_NO_OPENMP=0
"${CC}" "${CFLAGS}" -fopenmp test-omp.c || R_NO_OPENMP=1
```

I didn't know about these other flags that you've mentioned, but seems like a good idea to try for even more portability! If you open an issue listing those other flags (with the links you've included here) and and which compilers they are recognized by, I'd be happy to work on adding and testing that support.

Comment (StrikerRUS) [reply R-package/configure:1828]: > If you open an issue ...

Sure, just created #4537.

**Pull request merged**
